### ISSUE TITLE
InterruptExecutor doesn't work in rp2040 multicore.

### ISSUE THREAD

**Issue by Dirbaio** [embassy-rs/embassy#1634]
InterruptExecutor works by storing the interrupt number in the executor's Pender. So when a task is woken, the executor's interrupt number is pended in the NVIC. This causes the executor to get polled, which will poll the task.

The issue is if we have 2 cores, we actually have 2 NVICs. So:
- The executor is started on core0
- The executor's irq is enabled in core0's NVIC
- a waker gets sent from core0 to core1
- the waker is woken from core1
- the irq gets pended in core1's NVIC
- which does nothing because it's not enabled in core1's NVIC, only core0's.

The fix is:
- embassy-executor should get a "assume single core" Cargo feature or something, so it can choose to not expose `InterruptExecutor` in multicore.
- embassy-rp could expose a RP2040-specific InterruptExecutor, with some extra care to store "core num + irq num" in Pender, and make cross-core interrupts work. Maybe through FIFO?

Ideally it should also be possible to use the same irq number in the 2 cores simultaneously, for different InterruptExecutor instances. This requires the different cores to have different vector tables though, which `cortex-m-rt` doesn't support.

**Comment by kaspar030** [comment]
Could we encode the desired nvic(s) in the irq number, and then dispatch somehow when pending? there's only 32 irqs, 64 for both nvic, so plenty of bits left. :)

**Comment by stappersg** [comment]
_(-: Feel free to ignore this comment :-)_

> The fix is:
> 
>     * embassy-executor should get a "assume single core" Cargo feature or something, so it can choose to not expose `InterruptExecutor` in multicore.
> 
>     * embassy-rp could expose a RP2040-specific InterruptExecutor, with some extra care to store "core num + irq num" in Pender, and make cross-core interrupts work. Maybe through FIFO?

Having _multi core support_ in `embassy-rp` will restrict _embassy multi core support_ to RP2040 devices,  will hinder embassy supporting other multi core devices.

Giving `embassy-executor` feature like _core count_ seems to me the better long term option.    

> Could we encode the desired nvic(s) in the irq number, and then dispatch somehow when pending?

That sounds like _only ARM has multi core  devices_.  I do like to see that Embassy could support ESP32 (dual core)  and  multi core RISC-V devices.

**Comment by Dirbaio** [comment]
> Having multi core support in embassy-rp will restrict embassy multi core support to RP2040 devices, will hinder embassy supporting other multi core devices.

No, this doesn't prevent anyone else from writing multicore support for other chips in other chip-specific crates.

It is actually impossible to add "ARM multicore support" to `embassy-executor`. ARM doesn't define how multi-core communication works, each vendor invents their own IPC/FIFO peripherals. So to write a multicore executor you *must* use vendor-specific stuff. It can't go in `embassy-executor`.

**Comment by stappersg** [comment]
On Sun, Aug 13, 2023 at 12:42:36PM -0700, Dario Nieuwenhuis wrote:
> > Having multi core support in embassy-rp will restrict embassy multi core support to RP2040 devices, will hinder embassy supporting other multi core devices.
> 
> No, this doesn't prevent anyone else from writing multicore support
> for other chips in other chip-specific crates.
> 
> It is actually impossible to add "ARM multicore support" to
> `embassy-executor`. ARM doesn't define how multi-core communication
> works, each vendor invents their own IPC/FIFO peripherals. So to write
> a multicore executor you *must* use vendor-specific stuff. It can't
> go in `embassy-executor`.


OK, sounds good.  Thanks for sharing.

 
Groeten
Geert Stappers
-- 
Silence is hard to parse

**Comment by bugadani** [comment]
> It is actually impossible to add "ARM multicore support" to embassy-executor

At this point I believe the executors in embassy-executor should explicitly be declared single-core only.

- ESP32 Xtensa needs peripheral trickery to pend an interrupt across cores
- I don't know anything about RISC-V
- As per this issue, cortex-m seems to have the same issue as Xtensa
- I discount std and wasm

In ESP-land we have 4 specific interrupts, implemented outside the CPU cores, that are software pendable from either core, and can be handled by both cores. From skimming the datasheet it looks like the PIO could be used to emulate similar functionality on the RP without touching the FIFOs inside.

It should also be noted that executors can be implemented outside embassy-executor without disadvantage. It's also not very difficult to build an executor on top of `raw::Executor`, and embassy-executor could even provide building blocks to make chip-specific implementations minimal. This makes me wonder if embassy-executor actually needs to provide architecture-specific implementations.

**Comment by mschnell1** [comment]
Am I correct assuming that this issue only exists, if Embassy is using both cores, but not if any executor only is started on one core and the other core does it's thing (including interrupting) without Embassy involved ? 
In fact I did test the multiprio.rs example from \embassy\examples\rp\src\bin\, which uses the InterruptExecutor and seems to work fine.

**Comment by Dirbaio** [comment]
what's broken is 
- InterruptExecutor running in core A
- Task `foo` running in the InterruptExecutor
- Core B wakes a waker for task `foo`.

Note this doesn't need running any executor in core B, sending a waker from core A to core B and waking it there is enough to trigger the bug.

If you manually make sure to never ever wake an InterruptExecutor waker from the other core you might be able to avoid hitting the bug, but this is not something i'd recommend for using in production.

**Comment by mschnell1** [comment]
Yep.

We do not want to do something similar to SMP here :) 

The way to go with the user software is using the chip's hardware (Inter Processor FiFos / Mailboxes) to wake the core that runs the Executor.

Should be fine with most applications.

**Comment by Dirbaio** [comment]
yes

**Comment by sourcebox** [comment]
Solved by https://github.com/embassy-rs/embassy/pull/5409 so this issue can be closed.
